### DISCUSSION TITLE
Is there some manual way to add my own snippets to your plugin, or now I need to clone it and edit by hands?

### DISCUSSION THREAD

**Discussion by daniilrozanov** [iurimateus/luasnip-latex-snippets.nvim#3] [Q&A] [answered]
I think this plugin gives pretty good capabilities to add custom snippets

**Comment by iurimateus** [comment]
To remove a snippet, you can use the luasnip API. I think you might be able to mutate the tables that are required on init.lua, but it's a bit hacky.

You can override a snippet by adding another with the same trigger, but with a higher priority. Most have default prio of 0.

Example, overriding `mk` to `$$`

```lua
local ls = require("luasnip")
local conds = require("luasnip.extras.expand_conditions")
 
local utils = require("luasnip-latex-snippets.util.utils")

local not_math = utils.not_math() -- pass true if using Treesitter

local snip = ls.parser.parse_snippet({ trig = "mk", name = "Math" }, "$ ${1:${TM_SELECTED_TEXT}} $$0")

snip.condition = not_math
snip.priority = 10 -- anything greater than 0.

ls.add_snippets("tex", { snip }, {
    type = "autosnippets",
})
```


Do you have an interface in mind?

**Reply by daniilrozanov** [reply]
Thank you, this is what I needed, although I want to only add my own snippets and earlier I had some troubles with context conditions of snippets. And because conditions in this plugin work good, I can add more snippets not only for math environment.

**Reply by kiryph** [reply]
Can one specify the file where the given code snippet can be put?
Also considering lazy loading, e.g. by packer.nvim, so that the latex snippets are still only loaded for tex files.

Putting it into `~/.config/nvim/after/ftplugin/tex.lua` complains that `"luasnip-latex-snippets.util.utils"` cannot be found. I would have assumed that this file is sourced after ftplugins. But apparently not. For the record: the default snippets are loaded.

Also giving a hacky solution to remove a snippet would be helpful.

 And can one fix the code to be valid lua (lua is not javascript).

**Reply by kiryph** [reply]
I have added the snippet to the config function of packer.nvim. Do not to forget to run `:PackerCompile` (https://github.com/wbthomason/packer.nvim#compiling-lazy-loaders).

```lua
  use {
    'iurimateus/luasnip-latex-snippets.nvim',
    requires = { 'L3MON4D3/LuaSnip', 'lervag/vimtex' },
    config = function()
      require'luasnip-latex-snippets'.setup()

      -- Modify existing snippets using higher priority and util function from luasnip-latex-snippets
      local ls = require("luasnip")
      local s = ls.snippet
      local i = ls.insert_node
      local fmta = require("luasnip.extras.fmt").fmta
      local utils = require("luasnip-latex-snippets.util.utils")
      local not_math = utils.with_opts(utils.not_math, false) -- when using treesitter, change false to true

      ls.add_snippets("tex", {
        s({trig = "mk", snippetType="autosnippet", priority=100},
          fmta("$<>$<>", { i(1), i(2), }), {condition = not_math}
        ),
        -- ...
      })
    end,
    ft = "tex",
  }
```

**Reply by iurimateus** [reply]
> Also giving a hacky solution to remove a snippet would be helpful.

This is what I use for friendly-snippets:

```lua
local function dups_tex(s)
  local description = type(s.dscr) == "table" and s.dscr[1] or s.dscr
  return
    (
      s.name == "Bold Text"
      or s.name == "Italic Text"
      or s.name == "Math"
      or s.name == "Sub Section"
      or s.name == "Sub Sub Section"
    )
      or s.trigger == "item"
      or s.trigger:match "mathinline$"
      or ((s.name == "Theorem" or s.name == "Proof") and description and #description < 30)
end

local license_or_copyright = function(s)
  local description = type(s.dscr) == "table" and s.dscr[1] or s.dscr
  return description and description:match "License" or s.name == "copyright"
end

local function remove_s(s, fn)
  if s and fn and fn(s) then
    s:invalidate()
    return true
  end
end

-- Remove 'License' and 'copyright' snippets from friendly-snippets
local _remove = function(bufnr, ft)
  if bufnr == 0 or bufnr == nil then
    bufnr = vim.api.nvim_get_current_buf()
  end

  ft = ft or vim.bo[bufnr].filetype

  -- https://github.com/rafamadriz/friendly-snippets/blob/main/snippets/global.json
  local all_snippets = ls.get_snippets "all"
  local ft_snippets = ls.get_snippets(ft)

  for _, t in ipairs { all_snippets, ft_snippets } do
    for _, s in pairs(t) do
      remove_s(s, license_or_copyright)
    end
  end

  if ft == "tex" then
    for _, s in pairs(ft_snippets) do
      remove_s(s, dups_tex)
    end
  end
end

local known = {}
local g = vim.api.nvim_create_augroup("__remove_ls_snippets", {})
vim.api.nvim_create_autocmd("User", {
  group = g,
  pattern = "LuasnipSnippetsAdded",
  callback = function(args)
    local ft = vim.bo[args.buf].filetype or "empty"
    if not known[ft] then
      known[ft] = vim.loop.new_timer()
    end

    known[ft]:start(
      500,
      0,
      vim.schedule_wrap(function()
        _remove(args.buf, ft)
      end)
    )
  end,
})

vim.api.nvim_create_user_command("RemoveLicenseSnippets", function()
  _remove()
end, {})

_remove()
```

Timers (`known`) are leaking though.

> 
> And can one fix the code to be valid lua (lua is not javascript).

?

**Reply by kiryph** [reply]
> ?

I was referring to `var snip` and comments introduced with `//` which is not valid lua.

**Reply by iurimateus** [reply]
> I was referring to `var snip` and comments introduced with `//` which is not valid lua.

Thanks, fixed it.

**Reply by iurimateus** [reply, marked as answer]
update:

```lua
local ls = require("luasnip")
local utils = require("luasnip-latex-snippets.util.utils")
local is_math = utils.with_opts(utils.is_math, true) -- true to use treesitter
local not_math = utils.with_opts(utils.not_math, true) -- true to use treesitter

-- set a higher priority (defaults to 0 for most snippets)
local snip = ls.parser.parse_snippet(
  { trig = "mk", name = "Math", condition = utils.pipe({ not_math }), priority = 10 },
  "$ ${1:${TM_SELECTED_TEXT}} $$0"
)

ls.add_snippets("tex", { snip }, {
  type = "autosnippets",
})
```

**Reply by velloxen** [reply]
Hey, I've was trying to get this to work and was having issues (something about trying to call upvalue 'condition'). I figured out that LuaSnip expects the `[condition](https://github.com/L3MON4D3/LuaSnip/blob/master/DOC.md#snippets)` here to be a function `fn(line_to_cursor, matched_trigger, captures) -> bool` except `utils.not_math()` here resolves to a `bool` (and there seems to be similar problems with the other functions in `utils`). Should these be returning a function `() -> bool`? Or am I missing something?

**Reply by iurimateus** [reply]
Sorry for the late reply. You're correct, I've updated the example.
